pkg/reconciler/revision/resources/deploy.go

```diff
@@ -157,44 +157,64 @@ func makePodSpec(rev *v1.Revision, loggingConfig *logging.Config, tracingConfig
 }
 
 // BuildUserContainer makes a container from the Revision template.
-func BuildUserContainer(rev *v1.Revision) *corev1.Container {
-	userContainer := rev.Spec.GetContainer().DeepCopy()
+func BuildUserContainer(rev *v1.Revision) []corev1.Container {
+	containers := []corev1.Container{}
+
+	for i := range rev.Spec.PodSpec.Containers {
+		if len(rev.Spec.PodSpec.Containers[i].Ports) != 0 || len(rev.Spec.PodSpec.Containers) == 1 {
+			servingContainer := makeServingContainer(rev.Spec.GetContainer().DeepCopy(), rev)
+			updateImage(rev, &servingContainer)
+			containers = append(containers, servingContainer)
+		} else {
+			multiContainer := makeContainer(rev.Spec.PodSpec.Containers[i].DeepCopy(), rev)
+			updateImage(rev, &multiContainer)
+			containers = append(containers, multiContainer)
+		}
+	}
+	return containers
+}
+
+func updateImage(rev *v1.Revision, container *corev1.Container) {
+	for _, digest := range rev.Status.ContainerStatuses {
+		if digest.Name == container.Name {
+			container.Image = digest.ImageDigest
+		}
+	}
+}
+
+func makeContainer(container *corev1.Container, rev *v1.Revision) corev1.Container {
```

Comment:
I think this can take value. The compiler should be smart enough to elide copying on return.

Reply:
[rev](https://github.com/knative/serving/pull/7801/files#diff-b8668ec090ebde0668133377ef4dc9f0R160) and [container](https://github.com/knative/serving/pull/7801/files#diff-b8668ec090ebde0668133377ef4dc9f0R160) from DeepCopy gives value as pointer that's the reason i am not able to use argument as values

Reply:
Yeah but you can call `makeContainer(*x)`, it doesn't matter much :)

Reply:
Modified Thank you